File: Behavioral/Lab20/C#/Entities/States/BuyableState.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Interfaces;

namespace C_.Entities.States
{
    public class BuyableState : EstateState
    {
        public BuyableState(RealEstate estate) : base(estate)
        {
            
        }
        public override void Rent(Customer customer, DateTime startDate, DateTime endDate)
        {
            Console.WriteLine($"Customer {customer.NameAndSurname} can't rent an estate");
        }
        public override void Sell(Customer customer)
        {
            Console.WriteLine($"Customer {customer.NameAndSurname} can't sell an estate");
            
        }
        public override void SetEstateToBuy(Customer customer, double price)
        {
            if(!_estate.IsAlreadyBought){
            Console.WriteLine($"Customer {customer.NameAndSurname} bought an estate with price (USD) {price}");
            _estate.IsAlreadyBought = true;    
            }
            else
                Console.WriteLine($"Customer {customer.NameAndSurname} can't buy an estate because it is already bought");
        }

    }
}

File: Behavioral/Lab20/Typescript/src/Entities/States/buyable_state.ts
import { Customer } from "../customer";
import { EstateState } from "./estate_state";

export class BuyableState extends EstateState
{
    public rent(customer: Customer, startDate: Date, endDate: Date): void {
        console.log(`Customer ${customer.nameAndSurname} can't rent an estate`);
    }
    public sell(customer: Customer): void {
        console.log(`Customer ${customer.nameAndSurname} can't sell an estate`);
    }
    public setEstateToBuy(customer: Customer, price: number): void {
        if (!this._estate.isAlreadyBought) {
            console.log(`Customer ${customer.nameAndSurname} bought an estate with price (USD) ${price}`);
            this._estate.isAlreadyBought = true;
        } else {
            console.log(`Customer ${customer.nameAndSurname} can't buy an estate because it is already bought`);
        }
    }
}
